src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+			}
+		}
+
+		$this->fully_supported_input = 0 === count( $stack );
+
+		return $this->fully_supported_input;
+	}
+
+	public function next_tag( $query = null ) {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		if ( 0 < count( $this->open_elements ) ) {
+			$element = WP_HTML_Spec::element_info( end( $this->open_elements ) );
+			// @TODO: Handle self-closing HTML foreign elements: must convey self-closing flag on stack.
+			if ( $element::is_void ) {
+				array_pop( $this->open_elements );
+			}
+		}
+
+		if ( false === parent::next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		$element = WP_HTML_Spec::element_info( $tag_name );
+
+		$self_closes = $element::is_void || ( ! $element::is_html && $this->has_self_closing_flag() );
+		if ( $self_closes ) {
+			$this->open_elements[] = $tag_name;
+			return true;
+		}
+
+		if ( $this->is_tag_closer() ) {
+			array_pop( $this->open_elements );
+		} else {
+			$this->open_elements[] = $tag_name;
+		}
+
+		return true;
+	}
+
+	public function next_sibling() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( ! $this->is_tag_closer() && $current_depth === $starting_depth ) {
+				return true;
+			}
+
+			if ( ! $this->is_tag_closer() && $current_depth < $starting_depth ) {
+				return false;
+			}
+		}
+
+		return false;
+	}
+
+	public function next_child() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( ! $this->is_tag_closer() && $current_depth === $starting_depth + 1 ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	private function find_closing_tag() {
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( $this->is_tag_closer() && $current_depth < $starting_depth ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	public function get_inner_content() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		if ( ! $this->get_tag() || $this->is_tag_closer() ) {
+			return false;
+		}
+
+		$element = WP_HTML_Spec::element_info( $this->get_tag() );
+		if ( $element::is_void || ( ! $element::is_html && $this->has_self_closing_flag() ) ) {
+			return false;
+		}
+
+		// @TODO: Unique bookmark names
+		$this->set_bookmark( 'start' );
+		if ( ! $this->find_closing_tag() ) {
+			return false;
+		}
+		$this->set_bookmark( 'end' );
+
+		$start = $this->bookmarks['start']->end + 1;
```

Comment:
Here's one usage of the `end` bookmark property, but we could do without it by just storing this in a variable a few lines above. (I'm referring to the idea of ditching the `end` property)